### ISSUE TITLE
[sklearn#34626] `UnionFind.fast_find`'s path compression never touches the queried node — quadratic linkage labelling

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#307]
**Mirrored from [scikit-learn#34626](https://github.com/scikit-learn/scikit-learn/issues/34626)**

Original title: `UnionFind.fast_find`'s path compression never touches the queried node — quadratic linkage labelling
Original labels: Bug, Needs Triage
Created: 2026-08-01T10:04:27Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the bug and give evidence about its user-facing impact

`UnionFind.fast_find`'s path-compression loop never actually shortens the path of the node it was called on, so path compression is effectively disabled. `fast_find` degenerates to a plain, uncompressed find. Callers that build a dendrogram (e.g. `make_single_linkage`) allocate a new parent node per merge, so single-linkage/chaining inputs produce O(n)-deep union-find trees; with compression disabled, the repeated uncompressed finds sum to **O(n²)**. This is invisible on small inputs but dominates end-to-end runtime at scale (tens of seconds at n≈200k in a comparable reproduction — see the linked hdbscan report for measured numbers against the same code pattern).

**File:** `sklearn/cluster/_hierarchical_fast.pyx` — `UnionFind.fast_find`
**Also affects:** `sklearn/cluster/_hdbscan/_linkage.pyx::make_single_linkage`, which calls `UnionFind.fast_find` in its inner loop; and, within `sklearn/cluster/_agglomerative.py`, `AgglomerativeClustering(linkage="single")` / `FeatureAgglomeration(linkage="single")`, both of which route through `_hierarchical.single_linkage_label` / `_hierarchical._single_linkage_label` (the unconstrained and connectivity-constrained paths, respectively) in `linkage_tree()`. Other linkages (`ward`, `average`, `complete`) use a separate heap + `IntFloatDict` implementation and are **not** affected.
**Type:** performance (results are correct; time is quadratic)
**Related:** the same defect exists independently in the `hdbscan` package (`scikit-learn-contrib/hdbscan`, `hdbscan/_hdbscan_linkage.pyx`), which has its own vendored copy of this `UnionFind`. Bug report: [https://github.com/scikit-learn-contrib/hdbscan/issues/731](https://github.com/scikit-learn-contrib/hdbscan/issues/731).
**LLM usage**: found using Claude, reviewed manually.

### Steps/Code to Reproduce

```python
import numpy as np, time
from sklearn.cluster._hierarchical_fast import single_linkage_label

def build_L(K):
    rows = []
    w = 0
    for i in range(K - 1):
        rows.append((i, i + 1, w)); w += 1
    for i in range(K):
        rows.append((i, K + i, w)); w += 1
    return np.array(rows, dtype=np.float64)

prev_t = None
for K in [1000, 2000, 4000, 8000, 16000, 32000, 64000]:
    L = build_L(K)
    t0 = time.perf_counter()
    single_linkage_label(L)
    t1 = time.perf_counter()
    dt = t1 - t0
    ratio = dt / prev_t if prev_t else None
    print(f'K={K:6d}  n_samples={2*K:6d}  time={dt:.4f}s  ratio={ratio}')
    prev_t = d

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to https://github.com/godofecht/flow-scikit/issues/116 in the flow-scikit repository. Further library-specific triage and implementation will happen there.
